### PR TITLE
Decision forest feature

### PR DESCRIPTION
**Pull request by masdevas**
# Decision forest feature
Here is union of requests:
https://github.com/intel/daal/pull/333
https://github.com/intel/daal/pull/290
in addition: cpp example, java example, java interface

### CODE REVIEW

**Comment by masdevas** [comment]
CI with upgraded tests: https://intel-ci.intel.com/ea47e8b5-da78-f1b7-a176-8cdcd4b723a1

**Comment by masdevas** [comment]
make test

File: algorithms/kernel/dtrees/forest/classification/df_classification_model.cpp [hunk 1]
```diff
@@ -184,25 +187,27 @@ void ModelImpl::traverseDFS(size_t iTree, tree_utils::classification::TreeNodeVi
...
-    const DecisionTreeNode* aNode = (const DecisionTreeNode*)t.getArray();
-    const double *imp = getImpVals(iTree);
-    const int *nodeSamplesCount = getNodeSampleCount(iTree);
-    NodeIdxArray aCur;//nodes of current layer
-    NodeIdxArray aNext;//nodes of next layer
-    if(aNode)
+    if (iTree >= size()) return;
+    const DecisionTreeTable & t    = *at(iTree);
+    const DecisionTreeNode * aNode = (const DecisionTreeNode *)t.getArray();
+    const double * imp             = getImpVals(iTree);
```
Comment (ShvetsKS) [inline algorithms/kernel/dtrees/forest/classification/df_classification_model.cpp]: const double * const imp

File: algorithms/kernel/dtrees/forest/classification/df_classification_model_builder.cpp [hunk 2]
```diff
@@ -51,30 +51,76 @@ services::Status ModelBuilder::initialize(size_t nClasses, size_t nTrees)
...
+                                                                             res, modelImplRef._probTbl);
+}
+
+bool checkProba(const double * const proba, const size_t nClasses)
+{
+    double acc = 0.0;
+    const double epsilon = 10e-6;
+    for (size_t index = 0; index < nClasses; ++index)
+    {
+        if (proba[index] < 0.0 - epsilon || proba[index] > 1.0 + epsilon)
```
Comment (ShvetsKS) [inline algorithms/kernel/dtrees/forest/classification/df_classification_model_builder.cpp]: Do we really wand to allow negative values for probabilities?
Looks strange: -0.000001 will pass this check for example.

File: algorithms/kernel/dtrees/dtrees_model.cpp [hunk 3]
```diff
@@ -204,7 +207,35 @@ void setNode(DecisionTreeNode& node, int featureIndex, double response)
     node.featureValueOrResponse = response;
 }
 
-services::Status addSplitNodeInternal(data_management::DataCollectionPtr& serializationData,size_t treeId, size_t parentId, size_t position, size_t featureIndex, double featureValue, size_t& res)
+void setProbabilities(const size_t treeId, const size_t nodeId, const size_t response,
```
Comment (ShvetsKS) [inline algorithms/kernel/dtrees/dtrees_model.cpp]: why this function doesn`t return Status?

File: algorithms/kernel/dtrees/dtrees_model.cpp [hunk 4]
```diff
@@ -204,7 +207,35 @@ void setNode(DecisionTreeNode& node, int featureIndex, double response)
...
+            probOfTree[classIndex] = prob[classIndex];
+        }
+    }
+    else
+    {
+        for (size_t classIndex = 0; classIndex < nClasses; ++classIndex)
+        {
+            probOfTree[classIndex] = 0.0;
+        }
+        probOfTree[response] = 1.0;
```
Comment (ShvetsKS) [inline algorithms/kernel/dtrees/dtrees_model.cpp]: check should be added for 'response' LE than nClasses

File: algorithms/kernel/dtrees/forest/classification/df_classification_model_builder.cpp (hunk 2 above)
Comment (masdevas) [reply algorithms/kernel/dtrees/forest/classification/df_classification_model_builder.cpp]: Accumulated double values of probability can be a little bit more than 1 or less than 0.

File: include/algorithms/decision_forest/decision_forest_classification_predict_types.h [hunk 5]
```diff
@@ -106,9 +116,27 @@ class DAAL_EXPORT Input : public classifier::prediction::Input
...
+ * <a name="DAAL-STRUCT-ALGORITHMS__CLASSIFIER__PARAMETER"></a>
+ * \brief Class for the parameters of the Decision Forest classification algorithm
+ *
+ * \snippet decision_forest/decision_forest_classification_predict_types.h Parameter source code
+ */
+/* [interface1::Parameter source code] */
+struct DAAL_EXPORT Parameter : public daal::algorithms::classifier::interface2::Parameter
+{
+    Parameter(size_t nClasses, VotingMethod votingMethod = weighted) :
+        daal::algorithms::classifier::interface2::Parameter(nClasses),
```
Comment (ShvetsKS) [inline include/algorithms/decision_forest/decision_forest_classification_predict_types.h]: no explicit interface2 usage.

File: include/algorithms/decision_forest/decision_forest_classification_predict_types.h [hunk 6]
```diff
@@ -106,9 +116,27 @@ class DAAL_EXPORT Input : public classifier::prediction::Input
...
 };
 
+/**
+ * <a name="DAAL-STRUCT-ALGORITHMS__CLASSIFIER__PARAMETER"></a>
+ * \brief Class for the parameters of the Decision Forest classification algorithm
+ *
+ * \snippet decision_forest/decision_forest_classification_predict_types.h Parameter source code
+ */
+/* [interface1::Parameter source code] */
+struct DAAL_EXPORT Parameter : public daal::algorithms::classifier::interface2::Parameter
```
Comment (ShvetsKS) [inline include/algorithms/decision_forest/decision_forest_classification_predict_types.h]: no explicit interface2 usage.

File: algorithms/kernel/dtrees/dtrees_model.cpp (hunk 3 above)
Comment (masdevas) [reply algorithms/kernel/dtrees/dtrees_model.cpp]: All corner cases is processed outside this function

File: algorithms/kernel/dtrees/dtrees_model.cpp (hunk 4 above)
Comment (masdevas) [reply algorithms/kernel/dtrees/dtrees_model.cpp]: Check has been added in ModelBuilder::addLeafNodeInternal

**Comment by ShvetsKS** [comment]
Did you check dynamic Backward Compatibility?

**Comment by masdevas** [comment]
@ShvetsKS BC is checked

**Pull request merged**